test/kernel/extra_test.exs

```diff
@@ -52,4 +52,32 @@ defmodule Kernel.ExtraTest do
       assert_raise ArgumentError, fn -> TestModule.only_even!(3) end
     end
   end
+
+  describe "defunion" do
+    defmodule TestModule2 do
+      import Kernel.Extra
+
+      defunion :thing, [:foo, :bar, :baz]
+
+      def test_guard(x) when is_thing(x), do: true
+      def test_guard(_), do: false
+
+      def module_attr, do: @things
+    end
+
+    # Can't figure out how to test typespecs here, but maybe that's OK
+
+    test "defines a function returning a list of the given values" do
+      assert TestModule2.__union_things__() == [:foo, :bar, :baz]
+    end
+
+    test "defines a guard macro" do
+      assert TestModule2.test_guard(:foo) == true
+      assert TestModule2.test_guard(:qux) == false
```

Comment:
Since your macro is public, it might be worth testing it directly. ie. `TestModule2.is_thing(x)` instead of the extra surface area.

Reply:
I specifically wanted to test that it worked in a guard, which would require defining a function anyway - might as well define it in the module we're already defining